src/toil/job.py

```diff
@@ -59,246 +59,897 @@
-        self._cores = self._parseResource('cores', cores)
-        self._memory = self._parseResource('memory', memory)
-        self._disk = self._parseResource('disk', disk)
-        self._preemptable = preemptable
-        self._config = None
-
-    @property
-    def disk(self):
+    def __init__(self, promisingJob, recipientJob=None):
         """
-        The maximum number of bytes of disk the job will require to run.
+        :param toil.job.Job promisingJob: The job being asked for its return value.
+        :param toil.job.Job recipientJob: The job receiving the return value, if any.
         """
-        if self._disk is not None:
-            return self._disk
-        elif self._config is not None:
-            return self._config.defaultDisk
+        
+        self.promisingJob = promisingJob
+        self.recipientJob = recipientJob
+        
+        if recipientJob is None:
+            # Come up with a vaguer error message
+            super().__init__(f"Job {promisingJob.description} cannot promise its return value to a job that is not its successor")
         else:
-            raise AttributeError("Default value for 'disk' cannot be determined")
+            # Write a full error message
+            super().__init__(f"Job {promisingJob.description} cannot promise its return value to non-successor {recipientJob.description}")
+            
 
-    @property
-    def memory(self):
+class FakeID:
+    """
+    Placeholder for a job ID used by a JobDescription that has not yet been
+    registered with any JobStore.
+    
+    Needs to be held:
+        * By JobDescription objects to record normal relationships.
+        * By Jobs to key their connected-component registries and to record
+          predecessor relationships to facilitate EncapsulatedJob adding
+          itself as a child.
+        * By Services to tie back to their hosting jobs, so the service
+          tree can be built up from Service objects.
+    """
+    def __init__(self):
         """
-        The maximum number of bytes of memory the job will require to run.
+        Assign a unique temporary ID that won't collide with anything.
         """
-        if self._memory is not None:
-            return self._memory
-        elif self._config is not None:
-            return self._config.defaultMemory
-        else:
-            raise AttributeError("Default value for 'memory' cannot be determined")
+        self._value = uuid.uuid4()
+        
+    def __str__(self):
+        return self.__repr__()
+        
+    def __repr__(self):
+        return f'FakeID({self._value})'
+        
+    def __hash__(self):
+        return hash(self._value)
+        
+    def __eq__(self, other):
+        return isinstance(other, FakeID) and self._value == other._value
+        
+    def __ne__(self, other):
+        return not isinstance(other, FakeID) or self._value != other._value
 
-    @property
-    def cores(self):
+class ConfigClient:
```

Comment:
I suppose it might make sense to roll this into `Requirer`. I think I made `ConfigClient` first.

We can't initialize with a config in `__init__` because the flow we need to support is deserializing a JobDescription and then attaching it to the *current* config for this run of the workflow, whatever that is.

We could just make it a field that gets set, especially if `assignConfig` doesn't *actually* do any checking to catch multiple calls or configs having been serialized anything. But I wanted to make the semantics clear that when you ahve one of these things you *need* to give it a config before you really use it. The best I dea I came up with to do that was having a method.

I suppose I could change it around so that there's a property that inheriting classes use, and if `assignConfig` hasn't been called we throw when the property is read.